pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+        options.update(parameters)
+        tao.setOptionsPrefix(options.options_prefix)
+
+        tao.setFromOptions()
+
+        if tao.getType() in {PETSc.TAO.Type.LMVM, PETSc.TAO.Type.BLMVM}:
+            class InitialHessian:
+                pass
+
+            class InitialHessianPreconditioner:
+                def apply(self, pc, x, y):
+                    X = taoobjective.new_M_dual()
+                    from_petsc(x, X)
+                    assert len(taoobjective.reduced_functional.controls) == len(X)
+                    X = tuple(m._ad_convert_type(x, options=convert_options)
+                              for m, x in zip(taoobjective.reduced_functional.controls, X))
+                    to_petsc(y, X)
+
+            B_0_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                                  InitialHessian(), comm=comm)
+            B_0_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+            B_0_matrix.setUp()
+
+            B_0_matrix_pc = PETSc.PC().createPython(InitialHessianPreconditioner(),
+                                                    comm=comm)
+            B_0_matrix_pc.setOperators(B_0_matrix)
+            B_0_matrix_pc.setUp()
+
+            tao.setLMVMH0(B_0_matrix)
+            ksp = tao.getLMVMH0KSP()
+            ksp.setType(PETSc.KSP.Type.PREONLY)
+            ksp.setTolerances(rtol=0.0, atol=0.0, divtol=None, max_it=1)
+            ksp.setPC(B_0_matrix_pc)
+            ksp.setUp()
+        else:
+            B_0_matrix = None
+            B_0_matrix_pc = None
+
+        x = vec_interface.new_petsc()
+        tao.setSolution(x)
+        tao.setUp()
+
+        super().__init__(problem, parameters)
+        self._taoobjective = taoobjective
+        self._vec_interface = vec_interface
+        self._tao = tao
+        self._x = x
+
+        attach_destroy_finalizer(
+            self, tao, H_matrix, M_inv_matrix, B_0_matrix_pc, B_0_matrix, x)
+
+    @property
+    def taoobjective(self):
+        """The :class:`.TAOObjective` used for the optimization.
+        """
+
+        return self._taoobjective
+
+    @property
+    def tao(self):
+        """The :class:`petsc4py.PETSc.TAO` used for the optimization.
+        """
+
+        return self._tao
+
+    @property
+    def x(self):
+        """The :class:`petsc4py.PETSc.Vec` used to store the solution to the
+        optimization problem.
+        """
+
+        return self._x
+
+    def solve(self):
+        """Solve the optimization problem.
+
+        Returns:
+            OverloadedType or tuple[OverloadedType]: The solution.
+        """
+        M = tuple(
+            m.tape_value()._ad_copy()
+            for m in self.taoobjective.reduced_functional.controls)
+        self._vec_interface.to_petsc(self.x, M)
+        self.tao.solve()
+        self._vec_interface.from_petsc(self.x, M)
+        if self.tao.getConvergedReason() <= 0:
+            raise RuntimeError("Convergence failure")
```

Comment:
Is this the conventional exception for solver fails?

Reply:
Updated to use a `TAOConvergenceError` with a message similar to Firedrake linear solve failures.

pyadjoint doesn't seem to raise exceptions on for optimizer failures in the SciPy or ROL interfaces. I'm pretty sure that's unsafe for the SciPy interface, #166.